File: handlers/imageDropHandler.js
const { EmbedBuilder } = require('discord.js');
const fetch = (...args) => import('node-fetch').then(({ default: fetch }) => fetch(...args));
const fs = require('fs');
const path = require('path');
const sharp = require('sharp');
const tesseract = require('node-tesseract-ocr');
const Character = require('../models/Character');

module.exports = async (message) => {
    if (message.attachments.size > 0) {
        try {
            const tempDir = path.join(__dirname, '../temp');
            if (!fs.existsSync(tempDir)) {
                fs.mkdirSync(tempDir);
            }

            const attachment = message.attachments.first();
            const originalFilePath = path.join(tempDir, 'karuta_drop.png');

            const response = await fetch(attachment.url);
            const buffer = await response.buffer();
            fs.writeFileSync(originalFilePath, buffer);

            const crops = [
                { name: 'part1', left: 40, top: 60, width: 210, height: 50 },
                { name: 'part2', left: 310, top: 60, width: 205, height: 50 },
                { name: 'part3', left: 590, top: 60, width: 205, height: 50 }
            ];

            const resultsByPart = [];

            for (const crop of crops) {
                const cropFilePath = path.join(tempDir, `cropped_${crop.name}.png`);

                await sharp(originalFilePath)
                    .extract({ left: crop.left, top: crop.top, width: crop.width, height: crop.height })
                    .toFile(cropFilePath);

                const ocrData = await tesseract.recognize(cropFilePath, {
                    lang: 'eng',
                    oem: 3, // LSTM OCR engine
                    psm: 7 // Treat as a single text line
                });

                // Lọc bỏ các ký tự không mong muốn và xử lý văn bản
                const texts = ocrData
                    .split('\n')
                    .map(text => text.trim())
                    .filter(text => text && !/^\d{4,}$/.test(text))
                    .map(text => text.replace(/[)\»\|\}]/g, '')) // Loại bỏ ), », |, }
                    .filter(text => text); // Loại bỏ các chuỗi rỗng sau khi lọc

                resultsByPart.push(texts);

                fs.unlinkSync(cropFilePath);
            }

            fs.unlinkSync(originalFilePath);

            const lines = [];

            for (const texts of resultsByPart) {
                if (texts.length === 0) {
                    lines.push('None', '');
                    continue;
                }

                const character = texts.join(' ').trim(); // Gộp văn bản và loại bỏ khoảng trắng thừa

                // Tìm nhân vật với tên đã được làm sạch
                const results = await Character.find({ 
                    character: { $regex: `^${character}$`, $options: 'i' } // Tìm kiếm không phân biệt hoa thường
                });

                if (results.length > 0) {
                    for (const result of results) {
                        lines.push(`${character} (**${result.series}**): \`${result.wishlist} Wishlist\``);
                    }
                } else {
                    lines.push(`${character}: \`None\``);
                }
                lines.push('');
            }

            if (lines[lines.length - 1] === '') {
                lines.pop();
            }

            if (lines.length === 0) {
                console.log('⚠️ No valid characters found from OCR');
                return;
            }

            const embed = new EmbedBuilder()
                .setTitle('📊 Wishlist Stats')
                .setDescription(lines.join('\n'))
                .setColor('Blue');

            await message.channel.send({ embeds: [embed] });

        } catch (err) {
            console.error('❌ Error processing drop image:', err.stack);
            const tempDir = path.join(__dirname, '../temp');
            const originalFilePath = path.join(tempDir, 'karuta_drop.png');
            const cropFiles = ['part1', 'part2', 'part3'].map(name => path.join(tempDir, `cropped_${name}.png`));
            if (fs.existsSync(originalFilePath)) fs.unlinkSync(originalFilePath);
            cropFiles.forEach(file => {
                if (fs.existsSync(file)) fs.unlinkSync(file);
            });
        }
    }
};

File: handlers/image_drop_handler.py
import discord
import requests
import os
import shutil
from PIL import Image
import easyocr

async def handle_image_drop(message, db):
    if not message.attachments:
        return

    try:
        temp_dir = os.path.join(os.path.dirname(__file__), '../temp')
        os.makedirs(temp_dir, exist_ok=True)
        original_file_path = os.path.join(temp_dir, 'karuta_drop.png')

        # Tải hình ảnh
        attachment = message.attachments[0]
        response = requests.get(attachment.url, stream=True)
        with open(original_file_path, 'wb') as f:
            shutil.copyfileobj(response.raw, f)

        # Vùng cắt
        crops = [
            {'name': 'part1', 'left': 40, 'top': 60, 'width': 210, 'height': 50},
            {'name': 'part2', 'left': 310, 'top': 60, 'width': 205, 'height': 50},
            {'name': 'part3', 'left': 590, 'top': 60, 'width': 205, 'height': 50}
        ]

        reader = easyocr.Reader(['en'])
        results_by_part = []

        for crop in crops:
            crop_file_path = os.path.join(temp_dir, f'cropped_{crop["name"]}.png')
            with Image.open(original_file_path) as img:
                cropped = img.crop((crop['left'], crop['top'], crop['left'] + crop['width'], crop['top'] + crop['height']))
                cropped = cropped.convert('L').point(lambda x: 0 if x < 128 else 255, '1')  # Nhị phân hóa
                cropped.save(crop_file_path)

            result = reader.readtext(crop_file_path, detail=0, paragraph=True)
            texts = [text.strip() for text in result if text.strip()]
            texts = [text for text in texts if not text.isdigit() or len(text) < 4]
            texts = [text.replace(char, '') for text in texts for char in [')', '»', '|', '}', ';', '©', '~', '—']]
            texts = [text for text in texts if text]

            results_by_part.append(texts)
            os.unlink(crop_file_path)

        os.unlink(original_file_path)

        lines = []
        for texts in results_by_part:
            if not texts:
                lines.extend(['None', ''])
                continue

            character = ' '.join(texts).strip()
            # Truy cập trực tiếp collection 'characters' thay vì sử dụng class Character
            results = db.characters.find({'character': {'$regex': f'^{character}$', '$options': 'i'}})

            found = False
            for result in results:
                lines.append(f"{character} (**{result['series']}**): `{result['wishlist']} Wishlist`")
                found = True

            if not found:
                lines.append(f"{character}: `None`")
            lines.append('')

        if lines and lines[-1] == '':
            lines.pop()

        if not lines:
            print('⚠️ No valid characters found from OCR')
            return

        embed = discord.Embed(title='📊 Wishlist Stats', description='\n'.join(lines), color=0x0000FF)
        await message.channel.send(embed=embed)

    except Exception as err:
        print(f'❌ Error processing drop image: {err}')
        temp_dir = os.path.join(os.path.dirname(__file__), '../temp')
        original_file_path = os.path.join(temp_dir, 'karuta_drop.png')
        crop_files = [os.path.join(temp_dir, f'cropped_part{i}.png') for i in range(1, 4)]
        if os.path.exists(original_file_path):
            os.unlink(original_file_path)
        for file in crop_files:
            if os.path.exists(file):
                os.unlink(file)